src/atlas/field/detail/for_each.h

```diff
@@ -0,0 +1,275 @@
+ * nor does it submit to any jurisdiction.
+ */
+
+#pragma once
+
+#include <type_traits>
+#include <utility>
+#include <sstream>
+
+#include "atlas/field/Field.h"
+#include "atlas/array/helpers/ArrayForEach.h"
+#include "atlas/runtime/Exception.h"
+
+namespace atlas {
+namespace field {
+namespace detail {
+
+template <typename Function> // primary template
+struct function_traits : 
+    public function_traits<decltype(&std::remove_reference<Function>::type::operator())> { };
+
+template <typename ClassType, typename ReturnType, typename... Arguments>
+struct function_traits<ReturnType(ClassType::*)(Arguments...) const> : 
+    function_traits<ReturnType(*)(Arguments...)> { };
+
+template <typename ClassType, typename ReturnType, typename... Arguments>
+struct function_traits<ReturnType(ClassType::*)(Arguments...)> : 
+    function_traits<ReturnType(*)(Arguments...)> { };
+
+template <typename ReturnType, typename... Arguments>
+struct function_traits<ReturnType(*)(Arguments...)> {
+    typedef ReturnType result_type;
+
+    template <std::size_t Index>
+    using arg_t = typename std::tuple_element<
+        Index,
+        std::tuple<Arguments...>
+    >::type;
+
+    static constexpr std::size_t arity = sizeof...(Arguments);
+};
+
+template <typename Function>
+using first_argument = typename function_traits<Function>::template arg_t<0>;
+
+template <typename Function, typename Value>
+constexpr bool valid_value_function() {
+    return std::is_same_v<std::decay_t<first_argument<Function>>,std::decay_t<Value>>;
+}
+
+template <typename Function>
+using function_argument_data_type = typename std::decay_t<typename std::decay_t<first_argument<Function>>::value_type>;
+
+template <typename Function>
+constexpr idx_t function_argument_rank() {
+    return std::decay_t<first_argument<Function>>::rank();
+}
+
+template <typename Function, typename Value, int Rank>
+constexpr bool valid_column_function() {
+    using value_type = std::decay_t<Value>;
+    return function_argument_rank<Function>() == Rank && std::is_same_v<function_argument_data_type<Function>,value_type>;
+}
+
+template <typename Config,  typename Mask, typename Function, typename... Views, std::size_t... Dims>
+void for_each_value_masked_view(std::index_sequence<Dims...>, const Config& config, const Mask& mask, const Function& function, std::tuple<Views...>&& views ) {
+    array::helpers::ArrayForEach<Dims...>::apply(config,std::move(views),mask,function);
+}
+
+template <typename Value, int Rank, typename Config,  typename Mask, typename... Field, typename Function>
+void for_each_value_masked_rank(const Config& config, const Mask& mask, std::tuple<Field...>&& fields, const Function& function ) {
+    constexpr auto num_fields = std::tuple_size_v<std::tuple<Field...>>;
+    constexpr auto dims = std::make_index_sequence<Rank>();
+    if constexpr (valid_value_function<Function,Value>())
+    {
+        // Surely this can be improved to be more generic w.r.t. num_fields
```

Comment:
Surely we can!

```
template <int FieldIdx, int NumFields, typename Value, int Rank, typename... Field>
auto make_view_tuple(std::tuple<Field...>&& fields) {

    if constexpr (FieldIdx < NumFields) {
        return std::tuple_cat(std::make_tuple(
            std::move(array::make_view<Value, Rank>(std::get<FieldIdx>(fields)))),
            make_view_tuple<FieldIdx + 1, NumFields, Value, Rank>(std::move(fields)));
    } else {
        return std::make_tuple();
    }
}

template <typename Value, int Rank, typename Config,  typename Mask, typename... Field, typename Function>
void for_each_value_masked_rank(const Config& config, const Mask& mask, std::tuple<Field...>&& fields, const Function& function ) {
    constexpr auto num_fields = std::tuple_size_v<std::tuple<Field...>>;
    constexpr auto dims = std::make_index_sequence<Rank>();
    if constexpr (valid_value_function<Function,Value>())
    {
        // Surely this can be improved to be more generic w.r.t. num_fields
        // We have the technology!
        if constexpr (num_fields <= 4) {
            for_each_value_masked_view(dims, config, mask, function, make_view_tuple<0, num_fields, Value, Rank>(std::move(fields)));
            return;
        }
        ATLAS_THROW_EXCEPTION("Unsupported number of arguments in function");
    }
}
```

Reply:
And here I thought we'd seen the last of tuple_cats 😉
Thanks for your suggestion!

Reply:
It's my new favourite thing!

I don't think we need to pass `NumFields` into  `make_view_tuple` function, we can just use the size of the fields tuple.

Also, if we want to be really clever, I think we can 
pass in `Function` as a template parameter and use your SFINAE kung fu to get the `Value` type from each of the functor args.

Reply:
One problem with the function_traits is that we cannot use `auto` arguments in a "Function" lambda. This is related to runtime dispatch of fields based on ranks and data types, and it cannot infer this from `auto`.
So one has to be explicit w.r.t. data types and ranks in a view, and has to match what is encoded in the `Field`

Reply:
I was trying to wrap my head around this. You'd basically end up with thousands of compiled loop functions if you tried to dispatch every template combination. I interpreted your non-generic lambdas as trying to sneak in the template arguments by stealth ;-)